### ISSUE TITLE
Finalizers for components are useless (broken)

### ISSUE THREAD

**Issue by garyb** [slamdata/purescript-halogen#272]
The way things are currently, I don't think there's a case where a `finalizer` will ever work when used for a component.

The finalizer is triggered when an element is removed from the DOM, but because this happens when a component has been removed from the state, this means it is not possible to route the query to the component that was just removed.

**Comment by cryogenian** [comment]
There is global state too. For example `purescript-halogen-echarts` uses `Ref` to handle state of `echart` instances, and `finalizer` prevents memory leaks.

**Comment by garyb** [comment]
Yeah, I encountered this working on `new-halogen`: whenever you delete an Ace cell an "Always fails" error occurs from `Aff`, due to halogen being unable to route the `Quit` query back to the component so as to disable the autocomplete stuff... so there will be memory leaks all over the place right now :disappointed:

**Comment by natefaubion** [comment]
The reason why https://github.com/slamdata/purescript-halogen/blob/master/examples/lifecycle/src/Main.purs#L40-L41 works is because the lifecycle events are routed to the parent component.

I don't see how you could have proper finalization without decoupling lifecycles from virtual-dom (putting them in the component spec) and making it part of the render logic.

**Comment by garyb** [comment]
Yeah, I'm thinking that'll be the solution. I'd prefer initializers/finalizers to be associated with entire components rather than random HTML element(s) anyway too. Unless there are downsides to that I'm not seeing?

**Comment by jdegoes** [comment]
Perhaps component acquiring something like this would work:

```purescript
initialize :: Maybe (f Unit),
finalize :: Maybe (f Unit)
...
```

i.e. these are the commands to generate when the component is created and destroyed, respectively.

The installation combinators need the ability to run them on creation / destruction based on the placeholders, which I think should be possible.

**Comment by garyb** [comment]
Yeah, I'm already on it :smile:. Although `Maybe (HTMLElement -> f Unit)`, as we often need an element in these situations to perform the initialization of 3rd party components.

**Comment by natefaubion** [comment]
> Unless there are downsides to that I'm not seeing?

I don't see any, as this is how React works as well. Not quite sure how you'll get a hold of the HTMLElement from virtual-dom though.

**Comment by garyb** [comment]
I've spent some time looking into this now, and yeah there are a few tricky points. I don't really have the time to fix it right now, as the change is pretty big, so if anyone else wants to have a stab at it I'd greatly appreciate it!

I'd like to alter `Component`'s type to this:

``` purescript
newtype Component s f g = Component
  { render :: State s (HTML Void (f Unit))
  , eval :: Eval f s f g
  , initialize :: Maybe (HTMLElement -> f Unit)
  , finalize :: Maybe (f Unit)
  }
```

@natefaubion is right in saying that it's awkward to get hold of the `HTMLElement`, so by only requiring it for the initializer here it's going to save a little pain in the implementation, at the expense of components having to store that `HTMLElement` in their state if they require access to it to finalize themselves.

---

Determining when to execute the finalizer for a component can be done by diffing the list of keys in the `children` dictionary [after this line](https://github.com/slamdata/purescript-halogen/blob/a896ef5e55240de81b30e077c6fe3980ec0340aa/src/Halogen/Component.purs#L261). The "new" map of children is available from the the `State` monad we're operating in there, and the old state value is available as the `st` value in scope. 

The problem then is... we're only in the state monad, these actions somehow need to be executed. Maybe we could use the state monad writer-style and accumulate a list of finalizers that need to be executed, and then in the driver we can handle them post-render... I think we might need a `Ref` to do that, as we'd want all these things to end up in one place rather than being distributed throughout the component tree. Also, if a component is removed that has children where the finalizers need to run, etc. then we need to handle that. Since that component is gone, we can't keep track of the finalizers that need running there, we need to push them into the ref so even if all the components are deleted we still have somewhere that is tracking all of the pending finalizers.

---

Initializers are also tricky because of the `HTMLElement` thing, which means I think we still need to rely on `virtual-dom` as a way of inserting these, using the existing hook based approach. This means we need to have a consistent top level DOM element for any component that we want to initialize, so I think the approach will have to be to provide a component constructor that requires a wrapper element that we can then embed into the rendering to ensure we don't change the component from rendering a `div` to a `p` or something, as doing so would cause the initializer to run again. I'm not exactly sure what that would look like, but I'm suggesting components that use initializer/finalizer would be constructed something like this:

``` purescript
myComponent =
  lifecycleComponent
    [ P.class_ "my-component" ] 
    someRender 
    someEval 
    (Just someInitializer) 
    (Just someFinalizer)
```

Where `lifecycleComponent` is a new `Component` smart constructor. There'd need to be a corresponding `lifecycleParentComponent` with the appropriate types too. The idea being here that a consistent `div` is _always_ rendered for `lifecycleComponent`s, but some customization is provided by allowing an array of properties to be passed though (mainly thinking for class/id purposes). We can then hang our `initializer` off this div, safe in the knowledge that the only way it will fire is if the component is rendered in the DOM, and the div's content can be populated by the usual `render`.

**Comment by natefaubion** [comment]
I think that having initializers and finalizers be `f Unit` actually results in actions that are "too big". They should be a) synchronous and b) local. If you allow arbitrary evaluation it brings in a lot of questions like can parents peek in on children initialization/finalization (this seems weird)? Can you update state in a finalizer? What does it mean to have initializers/finalizers in Aff?

I personally think that an initializer should only need to do synchronous effects to setup a component and return a new state (something like `HTMLElement -> Eff eff State`), and a finalizer should only do synchronous effects to do any cleanup (something like `State -> Eff eff Unit`). I realize there should be some affordance for `subscribe` as well.

**Comment by garyb** [comment]
I agree about the async thing, and that peeking these would be weird (although if we adopt your new messaging scheme at some point that problem will go away, since there's no peeking -- I still haven't had time to play with that).

However... we try to avoid explicitly introducing `Eff` or `Aff` anywhere aside from the driver, as in theory, we'd like to be able to run "pure" components for testing purposes at some point. `EventSource`s currently scupper that, as it forces us into `MonadAff` at least for `g`, but again your messaging might rescue us in the future.

For initializers, perhaps `HTMLElement -> State -> g State` would be more suitable than `HTMLElement -> f Unit` (the initial state may be useful in there, if the component's `initialState` has config options included, for instance), as at least then we don't need to have weird potentially-peekable queries, while still avoiding explicit Eff/Aff. And then `State -> g Unit` for the finalizer.

I do agree that `g` is too powerful really, but I think it's the lesser evil than what happens if you try to integrate `Eff` into Halogen's types - I've just remembered the other reason we've avoided it as much as possible: introducing `Eff` means now you need to track the effect row everywhere too, unless you use an existential, `unsafeInterleaveEff`, and just say "anything goes" in these `Eff`s.

**Comment by garyb** [comment]
Oh, using `g` does mean we have to enable `subscribe` somehow still though...

**Comment by jdegoes** [comment]
I'll post my thoughts tomorrow.

**Comment by jdegoes** [comment]
@natefaubion `f Unit` is not too big, since it's merely the input to the query processor. That is, `f Unit` has no effects, it's merely an instruction that will be interpreted to some effect in the query processor.

@garyb I think this is an example of hitting the limitations of the highly magical virtual DOM layer. However, we're not at the point where we can replace that foundation or even know necessarily what to replace it with.

First off, the fact that the view is in `State` is somewhat of a hack. In fact, ideally creation of the view should be able to read from but not modify state. We needed the modification capability because we wanted to "do some stuff" with the install combinators, namely, automatically update the map from placeholder to child. 

This is another case where, during the rendering process, we want to "do some more stuff". So I'd suggest a similar approach will work here. Is it ideal? No, but I honestly think an ideal solution to this problem involves replacing the virtual dom magic with something that's "turtles all the way down".

So, let's say we add another capability to the renderer. What might that look like? 

The simplest approach is probably something like:

```purescript
newtype Component s f g = Component
  { render :: WriterT (List (f Unit)) (State s) (HTML Void (f Unit))
  , eval :: Eval f s f g
  , initialize :: Maybe (HTMLElement -> f Unit)
  , finalize :: Maybe (f Unit)
  }
```

In this formulation, the renderer may emit `f Unit`, which are instructions to the controller (either the installer combinators or the top level run UI) to feed various inputs into the query processor after the rendering process is complete. i.e. a rendering can not only modify a components state, but it may also schedule actions for execution after the rendering.

This is precisely the capability which gives you to run the initializers and finalizers inside the install combinators. In fact, the initializers and finalizers could, in theory, be removed from the definition of a component, and "layered in" via the combinators.

In fact, this capability is more powerful. If the rendering process can emit instructions for the query processor, then, in theory, the state need not be modified during rendering. Rather, a parent component may accept an enhanced (coproduct) input algebra, which has facilities for adding and removing children in response to DOM changes. What this means is that, again in theory, you can simplify a component to the following definition:

```purescript
newtype Component s f g = Component
  { render :: s -> Writer (List (f Unit)) (HTML Void (f Unit))
  , eval :: Eval f s f g
  }
```

The state modification goes away. In its place, a parent component's query processor accepts a new set of operations for adding and removing children in response to DOM changes (this change would remove complexity from the definition of a component and create stronger guarantees about rendering, but only by pushing complexity into the input algebra, so I'm not sure if it's worth it or not; TBD).

Generalizing rendering from "OK you can modify the state," to "you can tell the query processor to do some stuff," seems like a fairly modest change. And good until we figure out what a better foundation looks like on our way toward Halogen 1.0.

**Comment by natefaubion** [comment]
> @natefaubion f Unit is not too big, since it's merely the input to the query processor. That is, f Unit has no effects, it's merely an instruction that will be interpreted to some effect in the query processor.

This is true, since if you have to be in `g`, you might as well be in `f`.

> In fact, this capability is more powerful. If the rendering process can emit instructions for the query processor, then, in theory, the state need not be modified during rendering. Rather, a parent component may accept an enhanced (coproduct) input algebra, which has facilities for adding and removing children in response to DOM changes. What this means is that, again in theory, you can simplify a component to the following definition:
```purescript
newtype Component s f g = Component
  { render :: s -> Writer (List (f Unit)) (HTML Void (f Unit))
  , eval :: Eval f s f g
  }
```
I don't think you can quite reduce it to this since you still need access to real DOM elements for 3rd party libs. I think this is kind of neat, but practically speaking this means that each render potentially introduces an unbounded number of `Aff` binds which will probably become an issue if you need the guarantee that something is run within a single paint cycle.

**Comment by jdegoes** [comment]
> an unbounded number of Aff binds which will probably become an issue if you need the guarantee that something is run within a single paint cycle.

That's not exactly true. The render actually is totally synchronous and can't do anything. Parent components will schedule the post-render actions from their children, while the top-level run UI will be in charge of actually invoking the post-render actions after the rendering process is complete.

So you still have a strong guarantee of synchronous rendering. Made stronger still than the current guarantee, since with one of the above formulations, the rendering process cannot even modify the state (whereas in the current formulation, that happens all the time). What the rendering process can do, however, is produce a list of stuff that needs to be run — not run now, but run asynchronously, after the process is complete, and not necessarily in any fixed order (though preventing sequencing will require running in `Par` instead of `Aff`).

**Comment by natefaubion** [comment]
Sure, but I mean if you have a 3rd party lib that needs to be initialized synchronously so you don't get flickering. You can't guarantee that with Aff, and especially if you run it with Par.

**Comment by jdegoes** [comment]
> I don't think you can quite reduce it to this since you still need access to real DOM elements for 3rd party libs.

That's definitely true. But an install combinator currently has no way of locating the `HTMLElement` for an arbitrary `p`. Maybe this problem is best solved by extending `HTML` event listener to add one for a "where am I"-type event. Then install combinators can add these in the DOM, capture the HTML elements, and feed it into their query processors as appropriate.

**Comment by jdegoes** [comment]
> Sure, but I mean if you have a 3rd party lib that needs to be initialized synchronously so you don't get flickering. You can't guarantee that with Aff, and especially if you run it with Par.

`Aff` has a guarantee that a lifted synchronous action will be run synchronously. That said, it's possible such a guarantee can get lost during composition. That would be important to solve.

But you're right, definitely not acceptable for UI to flicker.

**Comment by garyb** [comment]
@jdegoes ah yes, I did muse the `Writer`+`State` approach but for some reason I apparently wasn't thinking straight when I wrote that wall of text up there. I had only intended to use it for handling `finalize`, due to the need to acquire an actual HTML element for `initialize`.

**Comment by garyb** [comment]
I'll see what I can do. I've already done a bunch of work that would be needed here, so I'll spend a little longer on it to try and get it into a more coherent shape and then open a PR so at least there's something concrete to discuss further.

**Comment by jdegoes** [comment]
@garyb No need to do this now if it will interfere with our Feb 1 target. Otherwise, knock yourself out. ;)

Unless, of course, we can get @natefaubion to volunteer to solve this problem. :laughing:

**Comment by garyb** [comment]
Yeah, I've still got that in mind! I'm not going to spend too much longer on it, but now my thinking has been unmuddied I might be able to make some good progress quickly.

And yeah, if I don't finish, maybe @natefaubion can take over from where I get up to :stuck_out_tongue_winking_eye:

**Comment by natefaubion** [comment]
> Aff has a guarantee that a lifted synchronous action will be run synchronously. That said, it's possible such a guarantee can get lost during composition. That would be important to solve.

Is this guarantee documented anywhere, BTW? Being potentially unbounded you still have a stack issue, and if you use `MonadRec` you introduce `later` which will definitely throw away any guarantees.

I can take a look at whatever @garyb comes up with, sure.

**Comment by natefaubion** [comment]
For the HTMLElements, would it make sense to follow React's lead? In React you annotate a DOM element with a `ref` name. In your controller you have access to the ref map, which is just an opaque data type. You can then ask React to give you the current element for that ref. Maybe we shouldn't give pass around the HTMLElement directly, and just make it part of HalogenF.

**Comment by garyb** [comment]
That seems reasonable. I'd considered integrating some notion of automatic ref assignments into the components as that gives us better behaviour with `virtual-dom` too: if you don't manually add `key`s when using `purescript-halogen-ace` sometimes weird things happen if you have more than one instance in the DOM at any one time. We could use the same value for both purposes.

**Comment by garyb** [comment]
I got basically nowhere with this in the end :wink:

I updated a bunch of signatures, and rendering run in `WriterT`, which produced a `List` of instructions to update the `State`, but slotting in the new stuff was going to take more thinking.

**Comment by natefaubion** [comment]
You mean the ref state? I think you probably have two options:
* Every component state gets wrapped in a `ComponentState` newtype that's like `InstalledState`, but with all the internal state that Halogen uses.
* It's kind of lame and gross, but since all components are globally addressable with a slot that's `Ord`, I don't think there's a reason you can't keep a big heap Map in the driver state of all of Halogen's internal bookkeeping. This would probably be very slow.

In any case, I think you could then make `ref` a hook that gets parameterized by a mutable `StrMap` in that internal state, and by the time the diffing process is done it should be populated. Then you can run the action's and they can access the elements.

With virtual-dom, I don't see how you can do this without loads of mutability at a distance.

**Comment by natefaubion** [comment]
The nice thing about the first option is that you can probably do it without changing user-facing signatures, since I don't think that type would need to be parameterized by anything other than the `s`, and its internals can be opaque. The app state is already mostly opaque since we don't export anyway to get the states out of `InstalledState`.

**Comment by natefaubion** [comment]
I actually sat down to look at the first option, and it obviously won't work because there's no access to state when rendering to virtual-dom (duh :P). Probably the easiest thing to do (if we are using the Writer approach for initializers/finalizers) is to just rename `initialize` to `ref`. This is actually in line with what React does now (string refs are effectively deprecated, you give ref a function that takes the component instance as an argument).

**Comment by natefaubion** [comment]
Or how about this, right now we think of "rendered" HTML as having a slot type of `Void`. But instead, you could think of it as having an arbitrary annotation on a sub tree of HTML. You could use this to propagate state to the final virtual-dom render, where it could collect things like initializers, finalizers, etc.

**Comment by jdegoes** [comment]
>Or how about this, right now we think of "rendered" HTML as having a slot type of Void. But instead, you could think of it as having an arbitrary annotation on a sub tree of HTML.

Suggests using a fixed-point encoding for `HTML` and `Cofree` for annotations. :smiling_imp:

**Comment by natefaubion** [comment]
> Suggests using a fixed-point encoding for HTML and Cofree for annotations.

I don't think we'd be doing a lot of redecoration (and all the subtrees need to be evaluated anyway), so I'm not sure how much the Cofree machinery would buy us. We do, however, a lot of mapping over HTML right now. Every time you nest components you have to a) map to install children b) map over child HTML to adjust event handlers. This results in a lot of traversing and garbage. If we wanted to get crazy, could we:

```purescript
newtype Biyoneda f a b = Biyoneda (forall c d. (a -> c) -> (b -> d) -> f c d)

type HTMLP = Biyoneda HTML
```

and get some sweet, sweet `bimap` fusion?

**Comment by natefaubion** [comment]
Does something like this work?
```purescript
data FinalizedF g s = FinalizedF s (s -> g Unit)

type Finalized g = Exists (FinalizedF g)

-- This may just need to be a very unsafe `StrMap HTMLElement` since we aren't
-- in Eff/Aff while we are rendering.
type RefMap = StrMap (Ref HTMLElement)

type RenderAnnotation f g =
  { thunk :: Boolean
  , initializer :: Maybe (f Unit)
  , finalizer :: Maybe (Finalized g)
  , refs :: RefMap
  , tree :: Rendered f g
  }

type Rendered f g = HTML (RenderAnnotation f g) (f Unit)

newtype InternalState s = InternalState { state :: s, refs :: RefMap }

newtype Component s f g = Component
  { render :: State (InternalState s) (Rendered f g)
  , eval :: Eval f (InternalState s) f g
  , initializer :: Maybe (f Unit)
  , finalizer :: Maybe (s -> g Unit)
  }
```

I don't really see how you could make a finalizer in `f`, as that would mean you'd have full access to `HalogenF`. Updating state, querying children, etc just doesn't make sense in a finalizer since its only useful for effectful cleanup.

**Comment by natefaubion** [comment]
I've got a preliminary PR working with lifecycles. Hopefully I can clean it up this weekend. A couple of notes:
* You can't just have a Writer with `f`, because you have the same issues with finalizers that we have now (the state doesn't exist in the tree anymore so it won't end up running)
* You should be able to use finalizers with `f` but you have to bundle up everything in an existential, and run it using an alternative driver that does not rerender or create subscriptions.
* So that means you should be able to use Writer with a sum of initializers and finalizers, where you run the initializers with the normal driver, and the finalizers with the alternative driver.
* I think it should be possible to bundle up the finalizer in such a way that `peek`s don't get invoked.

I think the `Writer` way works better. If you collect initializers/finalizers when rendering the vdom, then it just pushes that into `State` or `Writer`, which I don't like. The annotations are still useful to propagate things like thunks, so you really need both.

**Comment by natefaubion** [comment]
There's one major snag with finalizers. When you remove a component, it may have children itself, in which case you need to call all of the child finalizers first. There's no way to recursively collect finalizers though, because that would require you to open up a child's state during rendering (which obviously isn't possible since it's universally quantified). So the only option would be to call child finalizers within the parent finalizer. This should be possible but it means you are forced to query children 1 by 1, so you couldn't run finalizers in parallel. This is likely very bad if a finalizer does anything async :/

**Comment by natefaubion** [comment]
At the risk of rubber ducking too much, a crazy idea would be to always put finalizers on the HTML annotation. Then when you want to remove a child a component, you can traverse the `memo` tree for that component and collect child finalizers :) The annotated HTML just acts as a generic component tree that can be traversed. This is basically doing our own diffing outside of what virtual-dom does :O

**Comment by jdegoes** [comment]
To sketch my idea out a bit more, what happens is rendering loses the ability to modify state.

Instead, rendering may only produce a list of post-render actions (`f Unit`), via `Writer`.

Then a parent component supports a `Coproduct f (ParentComponentF p)`, where `ParentComponentF` looks something like this:

```purescript
data ParentComponentF p a
  = AddChild p a
  | RemoveChild p a
```

During rendering, the parent component render function looks over the new DOM to find `p`'s that have been added or removed, based on the state map. It translates these into actions in the parent component's input algebra, which is `Coproduct f (ParentComponentF p)`.

Then, the parent component interprets `right` actions and passes along `left` actions to the container component and / or the children components, as necessary. For `right` actions, the parent component leverages `StateF` to add and remove children in the state map.

This means that, there is essentially a one-step delay between when the DOM is modified to include new placeholders, and when those placeholders are actually substituted. This happens because the rendering process produces the actions to add / remove children, then the evaluator actually modifies the state, and then the rendering process is repeated to perform substitution. We could make this synchronous so there is no flickering, although it's obviously more work.

Note that a parent component must, on finalization, traverse the state map and run all the children finalizers. Since we provide the parent component combinator machinery, this probably only has to be done once there.

In addition, when interpreting `RemoveChild`, the parent component has the obligation to run finalizers. These finalizers may be defined in the child's `f` or `g`. If defined in `f`, then the parent can invoke the children interpreters to lift them to `g`. Or they can just be defined in `g` directly.

Thus, the final notion of a component looks something like:

```purescript
newtype Component s f g = Component
  { render :: s -> Writer (List (f Unit)) (HTML Void (f Unit))
  , eval :: Eval f s f g
  , initialize :: Maybe (HTMLElement -> Free g Unit)
  , finalize :: Maybe (Free g Unit)
  }
```

Plus or minus, depending on various decisions.

The drawback to this approach is that it complicates a parent component's input algebra, cluttering it with a `ParentComponentF` that will mostly be ignored by higher level components.

Also there may be some reason why the types can't line up, though it seems like they should.

**Comment by natefaubion** [comment]
I think the problem with parents dispatching child finalizers is as I tried to state above. You have to run each one after the other. So if g becomes Aff then one async finalizers can potentially hold up a whole chain of finalizers indefinitely. If you wanted to run them with Par then you'd have to fix g in the component definition, which we want to avoid. If you existentially seal the finalizers and collect them during rendering from the pruned trees, then you can avoid fixing g and still run them all with Par.

**Comment by jdegoes** [comment]
@natefaubion I wonder if coroutines could be a better target than either `Aff` or `Par`, assuming we had a `Par`-like means of composing them. Since what we really want to do is be able to combine things that eventually emit a single value, into something that emits a bunch of values at no specific time and in no specific order.

**Comment by natefaubion** [comment]
> Then a parent component supports a Coproduct f (ParentComponentF p), where ParentComponentF looks something like this:
``` purescript
data ParentComponentF p a
  = AddChild p a
  | RemoveChild p a
```

I don't think that's quite the full type. Since you you aren't updating the state during rendering, you have to propagate the `Component` definition _and_ state type (yet again) to this input algebra because it would have to be carried as part of `AddChild` so it can be inserted into the children map (imagine the type errors). I think that combined with that anyone who wants to query one of these components is going to have to go through yet another layer makes it somewhat unpalatable to me, even though it's a natural solution. I know some of the stuff I've suggested is a little weird (collecting finalizers off of trees), but it should result in no more external facing machinery.

**Comment by jdegoes** [comment]
> I know some of the stuff I've suggested is a little weird (collecting finalizers off of trees), but it should result in no more external facing machinery.

I definitely like no more external facing machinery as long as the result is sound and safe. :+1:

**Comment by natefaubion** [comment]
> I definitely like no more external facing machinery as long as the result is sound and safe. :+1:

Yes it should be. There's not really any funny business. The gist is that we take the HTML annotation idea, and change the signature of `render` to emit annotated HTML trees for components. When a component is rendered, it annotates itself with:

```purescript
data FinalizedF s f g = FinalizedF (Eval f s f g) s (f Unit)

foreign import data Finalized :: (* -> *) -> *

finalized :: forall s f g. Eval f s f g -> s -> f Unit -> Finalized g
finalized e s i = unsafeCoerce (FinalizedF e s i)

type ComponentTreeHTML f g = HTML (ComponentTree f g) (f Unit)

newtype ComponentTree f g = ComponentTree
  { html :: ComponentTreeHTML f g
  , finalizer :: Maybe (Finalized g)
  , thunk :: Boolean
  }
```

Which is just an existential package of `eval` with the current state and the designated input. The main driver/interpreter for `HalogenF` is already polymorphic in `s` and `f` (all it cares about is that `g` is `Aff`) so we aren't losing any important information that we'd need to actually run it.

We then change render from `State` to:
```purescript
type RenderM s f g = WriterT (Array (RenderHook f g)) (State s)

data RenderHook f g
  = PostRender (f Unit)
  | Finalized (Finalized g)
```

Parent components already carry around the cached trees of its children to avoid rerendering. When we compute which children have left the state, we can then look up its cached tree, and collect all the finalizers from the annotations, giving us a flat, ordered list of finalizers that we just emit using `Writer`.

Then in the driver, we can unpack these, apply the input, and run them using the supplied state. These can be run with `Par` or `forkAff` or whatever we decide on. I was saying we should use an alternative interpreter for finalizers, because we don't want them to cause rerendering of the main app and you don't want to create new subscriptions, since this part of the tree doesn't exist anymore.

This should be doable in a completely backwards compatible way since this is all just internal plumbing. `Component`s still take a `render` function of `s -> HTML Void (f Unit)` which gets coerced to `s -> HTML (ComponentTree f g) (f Unit)`. And parent components still take `SlotConstructor`s.

**Comment by natefaubion** [comment]
OK, so I got this working with an example https://github.com/natefaubion/purescript-halogen/commit/ae31e99f10de023babc6021df419645f588e82ad

I still need to workout `transform` and `interpret`, though it should be straightforward.

Otherwise, everything that currently works should work as is. I've added `componentSpec` and `parentComponentSpec` as a way to add lifecycles to components. It's still a bit messy, so please tear it apart, make suggestions, etc.

Some notes:
* it turns out, nothing is currently being stored in `memo` because `renderChild'` does not update the `memo` map. I obviously had to fix this.
* Because initializers get pushed through the main driver, they cause rerendering, though this should be easily fixed by just bypassing `driver` and and just evaling it directly.
* I went with `forkAff` for running everything as you still get synchronous behavior, but they won't hold up other inputs if they go async.

**Comment by garyb** [comment]
#280
